src/js_native_api_v8.cc

```diff
@@ -604,28 +620,72 @@ void Finalizer::ResetFinalizer() {
   finalize_hint_ = nullptr;
 }
 
-// Wrapper around v8impl::Persistent that implements reference counting.
-RefBase::RefBase(napi_env env,
-                 uint32_t initial_refcount,
-                 Ownership ownership,
-                 napi_finalize finalize_callback,
-                 void* finalize_data,
-                 void* finalize_hint)
+TrackedFinalizer::TrackedFinalizer(napi_env env,
+                                   napi_finalize finalize_callback,
+                                   void* finalize_data,
+                                   void* finalize_hint)
     : Finalizer(env, finalize_callback, finalize_data, finalize_hint),
-      refcount_(initial_refcount),
-      ownership_(ownership) {
+      RefTracker() {
   Link(finalize_callback == nullptr ? &env->reflist : &env->finalizing_reflist);
 }
 
-// When a RefBase is being deleted, it may have been queued to call its
+TrackedFinalizer* TrackedFinalizer::New(napi_env env,
+                                        napi_finalize finalize_callback,
+                                        void* finalize_data,
+                                        void* finalize_hint) {
+  return new TrackedFinalizer(
+      env, finalize_callback, finalize_data, finalize_hint);
+}
+
+// When a TrackedFinalizer is being deleted, it may have been queued to call its
 // finalizer.
-RefBase::~RefBase() {
+TrackedFinalizer::~TrackedFinalizer() {
   // Remove from the env's tracked list.
   Unlink();
   // Try to remove the finalizer from the scheduled second pass callback.
   env_->DequeueFinalizer(this);
 }
 
+void TrackedFinalizer::Finalize() {
+  FinalizeCore(/*deleteMe:*/ true);
+}
+
+void TrackedFinalizer::FinalizeCore(bool deleteMe) {
+  // Swap out the field finalize_callback so that it can not be accidentally
+  // called more than once.
+  napi_finalize finalize_callback = finalize_callback_;
+  void* finalize_data = finalize_data_;
+  void* finalize_hint = finalize_hint_;
+  ResetFinalizer();
+
+  // Either the RefBase is going to be deleted in the finalize_callback or not,
+  // it should be removed from the tracked list.
+  Unlink();
+  // 1. If the finalize_callback is present, it should either delete the
+  //    derived RefBase, or set ownership with Ownership::kRuntime.
+  // 2. If the finalizer is not present, the derived RefBase can be deleted
+  //    after the call.
+  if (finalize_callback != nullptr) {
+    env_->CallFinalizer(finalize_callback, finalize_data, finalize_hint);
+    // No access to `this` after finalize_callback is called.
```

Comment:
I have updated the comment. The code was adopted from the previously existed code. I cannot use the ownership variable because it is defined in the derived class. So, it was changed to use the `deleteMe` parameter.